pulser-core/pulser/sampler/samples.py

```diff
@@ -153,29 +166,96 @@ def modulate(
+        }
+        eom_samples = {
+            key: getattr(self, key).copy() for key in ("amp", "det")
+        }
+
+        if self.eom_blocks:
+            # Note: self.duration already includes the fall time
             eom_mask = np.zeros(self.duration, dtype=bool)
-            for start, end in self.eom_intervals:
-                end = min(end, self.duration)  # This is defensive
-                eom_mask[np.arange(start, end)] = True
+            # Extension of the EOM mask outside of the EOM interval
+            eom_mask_ext = eom_mask.copy()
+            eom_fall_time = 2 * cast(BaseEOM, channel_obj.eom_config).rise_time
+            for block in self.eom_blocks:
+                # If block.tf is None, uses the full duration as the tf
+                end = block.tf or self.duration
+                eom_mask[block.ti : end] = True
+                std_samples["amp"][block.ti : end] = 0
+                # For modulation purposes, the detuning on the standard
+                # samples is kept at 'detuning_off', which permits a smooth
+                # transition to/from the EOM modulated samples
+                std_samples["det"][block.ti : end] = block.detuning_off
+                # Extends EOM masks to include fall time
+                ext_end = end + eom_fall_time
+                eom_mask_ext[end:ext_end] = True
+
+            # We need 'eom_mask_ext' on its own, but we can already add it
+            # to the 'eom_mask'
+            eom_mask = eom_mask + eom_mask_ext
+
+            if block.tf is None:
+                # The sequence finishes in EOM mode, so 'end' was already
+                # including the fall time (unlike when it is disabled).
+                # For modulation, we make the detuning during the last
+                # fall time to be kept at 'detuning_off'
+                eom_samples["det"][-eom_fall_time:] = block.detuning_off
 
             for key in ("amp", "det"):
-                samples = getattr(self, key)
-                std = channel_obj.modulate(masked(samples, ~eom_mask))
-                eom = channel_obj.modulate(masked(samples, eom_mask), eom=True)
+                # First, we modulated the pre-filtered standard samples, then
+                # we mask them to include only the parts outside the EOM mask
+                # This ensures smooth transitions between EOM and STD samples
+                modulated_std = channel_obj.modulate(std_samples[key])
+                std = masked(modulated_std, ~eom_mask)
+
+                # At the end of an EOM block, the detuning is ramping back to
+                # detuning off by the EOM while detuning is being ramped down
+                # This equates to a ramp back to a modified detuning value,
+                # so we subsitute the detuning at the end of each block by the
+                # standard modulated detuning during the transition period
+                # Finally, the modified EOM samples are modulated
```

Comment:
When we disable the EOM mode, multiple things happen:
- The EOM configuration goes back to the OFF state, meaning that we will have a fast change towards `detuning_off` (exactly as we have in between EOM pulses)
- Both beams are turned off (which happens slowly, dictated by the standard modulation bandwidth)
- The detuning is also brought back to zero (during EOM mode, it is set such that ∂ = `detuning_on` - `lightshift_on` , such that we have `detuning_on` during the pulses as requested), and this transition is also slow.

So, we end up having these two competing effects: the detuning is ramping slowly back to 0 (both because the detuning we apply and the light-shift are ramping down), while it also tries to ramp quickly to `detuning_off`, so we end up having this fast ramp to a modified detuning (it doesn't reach `detuning_off` because at that point the detuning has already started going to 0).

Reply:
OK, I see. It's a bit more clear now. The explanation is likely too complex to put it in a comment but it might be worth extending a bit what you have now.

Reply:
Ok, will do!